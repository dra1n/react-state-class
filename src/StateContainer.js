import EventEmitter from 'events'

class StateContainer extends EventEmitter {
  update() {
    this.emit('change')
  }
}

export default StateContainer
