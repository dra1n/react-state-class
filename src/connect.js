import React, { Component } from 'react'

const connect = ({
  initState = () => ({}),
  stateHandler
}) => (mapStateToProps, mapStateHandlersToProps) => WrappedComponent => {
  return class StateContainer extends Component {
    constructor(props) {
      super(props)

      const StateHandler = stateHandler
      this.stateContainer = new StateHandler(initState(props))
    }

    componentDidMount() {
      this.stateContainer.on('change', () => {
        this.forceUpdate()
      })
    }

    componentWillUnmount() {
      this.stateContainer.removeAllListeners('change')
    }

    render() {
      const stateProps = mapStateToProps(
        this.stateContainer
      )

      const stateHandlerProps = mapStateHandlersToProps(
        this.stateContainer
      )

      return (
        <WrappedComponent
          {...this.props}
          {...stateProps}
          {...stateHandlerProps}
        />
      )
    }
  }
}

export default connect
