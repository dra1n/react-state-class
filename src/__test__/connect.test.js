/* globals mount, shallow */
/* eslint react/prop-types: 0 */

import React from 'react'
import renderer from 'react-test-renderer'
import connect from '../connect'
import StateContainer from '../StateContainer'

class UserMenuState extends StateContainer {
  constructor({authenticated}) {
    super()
    this.authenticated = authenticated
  }
  get menuOptions() {
    return this.authenticated ? 'Log Out' : 'Press authenticate'
  }
  authenticate() {
    this.authenticated = true
    this.update()
  }
}

const UserMenu = ({ menuOptions, authenticate }) => (
  <div>
    <div>
      { menuOptions }
    </div>
    <a onClick={authenticate}>
      Authenticate
    </a>
  </div>
)

const initState = props => ({ authenticated: props.authenticated })
const mapStateToProps = stateContainer => ({ menuOptions: stateContainer.menuOptions })
const mapStateHandlersToProps = stateContainer => ({
  authenticate: stateContainer.authenticate.bind(stateContainer)
})
const StatefulUserMenu = connect({
  initState: initState,
  stateHandler: UserMenuState
})(mapStateToProps, mapStateHandlersToProps)(UserMenu)


describe('connect', () => {
  it('passes props to the given component', () => {
    const tree = shallow(<StatefulUserMenu title='User Menu' />)
    expect(tree.prop('title')).toEqual('User Menu')
  })

  it('passes state container properties and methods to the given component', () => {
    const tree = shallow(<StatefulUserMenu />)
    expect(tree.prop('menuOptions')).toBeDefined()
    expect(tree.prop('authenticate')).toBeDefined()
  })

  describe('Stateful Component', () => {
    it('renders correctly', () => {
      let tree = renderer.create(<StatefulUserMenu authenticated={false} />).toJSON()
      expect(tree).toMatchSnapshot()

      tree = renderer.create(<StatefulUserMenu authenticated={true} />).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('reacts to changes inside state container', () => {
      const tree = mount(<StatefulUserMenu authenticated={false} />)
      tree.find('a').simulate('click')
      expect(tree).toMatchSnapshot()
    })
  })
})
