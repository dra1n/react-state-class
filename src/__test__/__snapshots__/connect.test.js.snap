// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect Stateful Component reacts to changes inside state container 1`] = `
<StateContainer
  authenticated={false}
>
  <UserMenu
    authenticate={[Function]}
    authenticated={false}
    menuOptions="Log Out"
  >
    <div>
      <div>
        Log Out
      </div>
      <a
        onClick={[Function]}
      >
        Authenticate
      </a>
    </div>
  </UserMenu>
</StateContainer>
`;

exports[`connect Stateful Component renders correctly 1`] = `
<div>
  <div>
    Press authenticate
  </div>
  <a
    onClick={[Function]}
  >
    Authenticate
  </a>
</div>
`;

exports[`connect Stateful Component renders correctly 2`] = `
<div>
  <div>
    Log Out
  </div>
  <a
    onClick={[Function]}
  >
    Authenticate
  </a>
</div>
`;
